src/pages/en/core-concepts/endpoints.md

```diff
@@ -0,0 +1,275 @@
+
+In that file, export a `get` function (optionally `async`) that returns an object with a `body`. In static mode, Astro will call this at build time to generate the corresponding file.
+
+```js
+// Example: src/pages/builtwith.json.ts
+// Outputs: /builtwith.json
+
+// File routes export a get() function, which gets called to generate the file.
+// Return an object with `body` to save the file contents in your final build.
+export async function get({params, request}) {
+  return {
+    body: JSON.stringify({
+      name: 'Astro',
+      url: 'https://astro.build/',
+    }),
+  };
+}
+```
+
+The return object can also have an `encoding` property. It can be any valid [`BufferEncoding`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bdd02508ddb5eebcf701fdb8ffd6e84eabf47885/types/node/buffer.d.ts#L169) accepted by node.js' `fs.writeFile` method. For example, to produce a binary png image:
+
+```ts title="src/pages/astro-logo.png.ts" {7}
+export async function get({ params, request }) => {
+  const response = await fetch("https://astro.build/assets/press/full-logo-light.png");
+  const buffer = Buffer.from(await response.arrayBuffer());
+  return {
+    body: buffer,
+    encoding: 'binary',
+  };
+```
+
+You can type your endpoint functions using the `APIRoute` type:
+
+```js
+import type { APIRoute } from 'astro';
+
+export const get: APIRoute = async function get ({params, request}) {
+...
+```
+
+As their only argument, endpoint functions recieve an object with two properties: `params` and `request`.
+
+### `params` and Dynamic routing
+
+Endpoints support the same [dynamic routing](/en/core-concepts/routing/#dynamic-routes) features that pages do. Name your file with a bracketed parameter name and export a [`getStaticPaths()` function](/en/reference/api-reference/#getstaticpaths). Then, you can access the parameter using the `params` property passed to the endpoint function:
+
+```ts title="src/pages/[id].json.ts"
+import type { APIRoute } from 'astro';
+
+const usernames = ["Sarah", "Chris", "Dan"]
+
+export const get: APIRoute = ({ params, request }) => {
+  const id = params.id;
+  return {
+    body: JSON.stringify({
+      name: usernames[id]
+    })
+  }
+};
+
+export function getStaticPaths () {
+    return [ 
+        { params: { id: "0"} },
+        { params: { id: "1"} },
+        { params: { id: "2"} },
+    ]
+}
+```
+
+This will generate three JSON endpoints at build time: `/api/1.json`, `/api/2.json`, `/api/3.json`. Dynamic routing with endpoints works the same as it does with pages, but because the endpoint is a function and not a component, [props](/en/reference/api-reference/#data-passing-with-props) aren't supported.
```

Comment:
```suggestion
This will generate three JSON files at build time: `/api/1.json`, `/api/2.json`, `/api/3.json`. Dynamic routing with endpoints works the same as it does with pages, but because the endpoint is a function and not a component, [props](/en/reference/api-reference/#data-passing-with-props) aren't supported.
```

Reply:
Also, for SSG mode we probably could support `props`, at least I don't see a reason why not. I don't think we excluded that on purpose.